storage/migration/migration.go

```diff
@@ -0,0 +1,293 @@
+	lgProgress func(int),
+	db *badger.DB,
+	jobs <-chan []byte, // each job is a prefix to iterate over
+	kvChan chan<- []KVPair, // channel to send key-value pairs to writer workers
+	batchSize int,
+) error {
+	for prefix := range jobs {
+		err := db.View(func(txn *badger.Txn) error {
+			defer lgProgress(1)
+
+			if ctx.Err() != nil {
+				return ctx.Err()
+			}
+
+			options := badger.DefaultIteratorOptions
+			options.Prefix = prefix
+			it := txn.NewIterator(options)
+			defer it.Close()
+
+			var (
+				kvBatch  []KVPair
+				currSize int
+			)
+
+			for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+				if ctx.Err() != nil {
+					return ctx.Err()
+				}
+
+				item := it.Item()
+				key := item.KeyCopy(nil)
+				val, err := item.ValueCopy(nil)
+				if err != nil {
+					return err
+				}
+
+				kvBatch = append(kvBatch, KVPair{Key: key, Value: val})
+				currSize += len(key) + len(val)
+
+				if currSize >= batchSize {
+					select {
+					case kvChan <- kvBatch:
+					case <-ctx.Done():
+						return ctx.Err()
+					}
+					kvBatch = nil
+					currSize = 0
+				}
+			}
+
+			if len(kvBatch) > 0 {
+				select {
+				case kvChan <- kvBatch:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+			}
+
+			return nil
+		})
+
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// writerWorker writes key-value pairs to PebbleDB in batches.
+func writerWorker(ctx context.Context, db *pebble.DB, kvChan <-chan []KVPair) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case kvGroup, ok := <-kvChan:
+			if !ok {
+				return nil
+			}
+			batch := db.NewBatch()
+			for _, kv := range kvGroup {
+				if err := batch.Set(kv.Key, kv.Value, nil); err != nil {
+					return fmt.Errorf("fail to set key %x: %w", kv.Key, err)
+				}
+			}
+
+			if err := batch.Commit(nil); err != nil {
+				return fmt.Errorf("fail to commit batch: %w", err)
+			}
+		}
+	}
+}
+
+// CopyFromBadgerToPebble migrates all key-value pairs from a BadgerDB instance to a PebbleDB instance.
+//
+// The migration is performed in parallel using a configurable number of reader and writer workers.
+// Reader workers iterate over the BadgerDB by sharded key prefixes (based on ReaderShardPrefixBytes)
+// and send key-value pairs to a shared channel. Writer workers consume from this channel and write
+// batched entries into PebbleDB.
+//
+// Configuration is provided via MigrationConfig:
+//   - BatchByteSize: maximum size in bytes for a single Pebble write batch.
+//   - ReaderWorkerCount: number of concurrent workers reading from Badger.
+//   - WriterWorkerCount: number of concurrent workers writing to Pebble.
+//   - ReaderShardPrefixBytes: number of bytes used to shard the keyspace for parallel iteration.
+//
+// The function blocks until all keys are migrated and written successfully.
+// It returns an error if any part of the process fails.
+func CopyFromBadgerToPebble(badgerDB *badger.DB, pebbleDB *pebble.DB, cfg MigrationConfig) error {
```

Comment:
This approach works, but I noticed it's quite slow, since inserting keys to pebble will often trigger compaction. 

Pebble actually provides another way to bulk insert key-value pairs by directly writing to sstables. So as long as the key-value pairs are sorted when writing to the sstables, it can avoid compaction, and therefore much faster. 

I'm going to try in a separate PR, but reuse the same test cases.

Reply:
@zhangchiqing 

If data migration is too slow, maybe try these Pebble settings and manual compaction:
- `Options.DisableWAL` to disable writing WAL files
- `Options.DisableAutomaticCompactions` to disable compaction during migration 
- After migration completes, use `Compact()` with the `parallelize` parameter set to `true` for manual compaction.